### ISSUE TITLE
Investigate unexpected search order behaviour 

### ISSUE THREAD

**Issue by lbucas** [hochschule-darmstadt/openartbrowser#85]
**Is your feature request related to a problem? Please describe.**
When searching for "Renaissance", results in many categories appear (as motif, genre & movement). The expected search order would put the movement on top, since it has the most artworks assigned to it. Strangely, the movement appears as last result right now. We should investigate if this a special case with "Renaissance" or it is a deeper-lying bug.

**Why do you want this feature? What goals should be achieved**
The behavior could be confusing for the user and should be fixed.

**Describe the solution you'd like**
High ranked results should appear higher in the search order.

**Describe acceptance criteria**
The search order works as intended. High ranked results appear first.

**Describe alternatives you've considered**
Leave it as it is.

**What effects does your proposed solution have**
Depending on where the problem originates, some changes to the search order algorithm in the frontend or the ranking algorithm in the backend are necessary.

**Additional context**
![image](https://user-images.githubusercontent.com/32192444/69801299-3307a000-11d7-11ea-9568-65f6625df3cd.png)

**Comment by BernhardHumm** [comment]
Please note: in the autosuggestions, the different categories (movement, genre, motif etc.) should not be mixed but should be grouped (as they are now). 

The change request relates to the sorting of the groups only. The groups shall be sorted according to the rank of the top-scoring element.

 

           Bernhard Humm

 

 

 

Von: Lucas Bergmann [mailto:notifications@github.com] 
Gesendet: Donnerstag, 28. November 2019 12:04
An: hochschule-darmstadt/openartbrowser <openartbrowser@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Betreff: [hochschule-darmstadt/openartbrowser] Investigate unexpected search order behaviour (#85)

 

Is your feature request related to a problem? Please describe.
When searching for "Renaissance", results in many categories appear (as motif, genre & movement). The expected search order would put the movement on top, since it has the most artworks assigned to it. Strangely, the movement appears as last result right now. We should investigate if this a special case with "Renaissance" or it is a deeper-lying bug.

Why do you want this feature? What goals should be achieved
The behavior could be confusing for the user and should be fixed.

Describe the solution you'd like
High ranked results should appear higher in the search order.

Describe acceptance criteria
The search order works as intended. High ranked results appear first.

Describe alternatives you've considered
Leave it as it is.

What effects does your proposed solution have
Depending on where the problem originates, some changes to the search order algorithm in the frontend or the ranking algorithm in the backend are necessary.

Additional context
 <https://user-images.githubusercontent.com/32192444/69801299-3307a000-11d7-11ea-9568-65f6625df3cd.png>

**Comment by xcaverxiii** [comment]
>The change request relates to the sorting of the groups only. The groups shall be sorted according to the rank of the top-scoring element.

So it is implemented today:

First of all Artworks will be shown every time at first group
```
entities.sort((a, b): any => {
     return (a.type === 'artwork' && b.type !== 'artwork') || a.type > b.type;
});
```
And then for all others it will be sortet by group name i quess
```
if (typeB < typeA) {
      return 1;
} else if (typeA < typeB) {
      return -1;
}
```

Code can be found in the search.component.ts

**Comment by BernhardHumm** [comment]
That the image group is displayed first is ok but not necessary.

The order of all (other) groups should be descending according to the ranking of the top-ranked entity (ranking includes relative rank of entity and various factors)

 

           Bernhard Humm

 

------------------------------------

 

 

Von: Xcaver [mailto:notifications@github.com] 
Gesendet: Mittwoch, 11. Dezember 2019 14:42
An: hochschule-darmstadt/openartbrowser <openartbrowser@noreply.github.com>
Cc: Humm, Bernhard, Prof. Dr. <bernhard.humm@h-da.de>; Comment <comment@noreply.github.com>
Betreff: Re: [hochschule-darmstadt/openartbrowser] Investigate unexpected search order behaviour (#85)

 

The change request relates to the sorting of the groups only. The groups shall be sorted according to the rank of the top-scoring element.

So it is implemented today:

First of all Artworks will be shown every time at first group

entities.sort((a, b): any => {
     return (a.type === 'artwork' && b.type !== 'artwork') || a.type > b.type;
});

And then for all others it will be sortet by group name i quess

if (typeB < typeA) {
      return 1;
} else if (typeA < typeB) {
      return -1;
}

Code can be found in the search.component.ts

**Comment by istthabel** [comment]
Search suggestions are corrected and look like this now for 'Renaissance':
![newSearch](https://user-images.githubusercontent.com/28310087/70673169-e796cc80-1c81-11ea-9329-ee4da7ac9fce.png)
Results are ranked by relative rank. Relative rank is doubled, if item label begins with the search term, and halved, if search term is found in the middle of a word. Type is only used for grouping afterwards. Groups start with type with highest modified rank

**Comment by istthabel** [comment]
fixed and merged #162
